vms/evm/subscriber.go

```diff
@@ -154,25 +159,39 @@ func (s *subscriber) ProcessFromHeight(height *big.Int) error {
 		return err
 	}
 
-	// Cap the number of blocks to process to MaxBlocksToProcess
-	toBlock := big.NewInt(0).SetUint64(latestBlock)
-	if height.Cmp(big.NewInt(0).Sub(toBlock, big.NewInt(MaxBlocksToProcess))) < 0 {
-		s.logger.Warn(
-			fmt.Sprintf("Requested to process too many blocks. Processing only the most recent %d blocks", MaxBlocksToProcess),
-			zap.String("requestedBlockHeight", height.String()),
-			zap.String("latestBlockHeight", toBlock.String()),
-			zap.String("chainID", s.chainID.String()),
-		)
-		height = big.NewInt(0).Add(toBlock, big.NewInt(-MaxBlocksToProcess))
+	bigLatestBlock := big.NewInt(0).SetUint64(latestBlock)
+
+	// helper function to determine the upper bound of a block range given
+	// the lower bound:
+	toBlock := func(fromBlock *big.Int) (toBlock *big.Int) {
+		toBlock = big.NewInt(0).Add(fromBlock, big.NewInt(MaxBlocksPerRequest))
+		if toBlock.Cmp(bigLatestBlock) > 0 {
+			toBlock.Set(bigLatestBlock)
+		}
+		return
 	}
 
-	// Filter logs from the latest processed block to the latest block
-	// Since initializationFilterQuery does not modify existing fields of warpFilterQuery,
-	// we can safely reuse warpFilterQuery with only a shallow copy
+	for fromBlock := big.NewInt(0).Set(height); fromBlock.Cmp(bigLatestBlock) <= 0; fromBlock.Add(toBlock(fromBlock), big.NewInt(1)) {
+		err = s.processBlockRange(ethClient, fromBlock, toBlock(fromBlock))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Filter logs from the latest processed block to the latest block
+// Since initializationFilterQuery does not modify existing fields of warpFilterQuery,
+// we can safely reuse warpFilterQuery with only a shallow copy
+func (s *subscriber) processBlockRange(
+	ethClient ethclient.Client,
+	fromBlock, toBlock *big.Int,
+) error {
 	initializationFilterQuery := interfaces.FilterQuery{
 		Topics:    warpFilterQuery.Topics,
 		Addresses: warpFilterQuery.Addresses,
-		FromBlock: height,
+		FromBlock: fromBlock,
 		ToBlock:   toBlock,
```

Comment:
I played around with some public ethereum nodes and found that calling `eth_getLogs` with a higher-than-latest `toBlock` simply gives logs to the latest block. (It doesn't error.) I suppose that in theory this means we might be able to simplify some of the logic here, but given that we're subscribing before catching up (see [here](https://github.com/ava-labs/awm-relayer/blob/main/relayer/relayer.go#L121-L130)), I'd suggest that we leave that clamping logic in place, so that we don't doubly queue any blocks.

Reply:
I played around with some public ethereum nodes and found that calling `eth_getLogs` with a higher-than-latest `toBlock` simply gives logs to the latest block. (It doesn't error.) I suppose that in theory this means we might be able to simplify some of the logic here, but given that we're subscribing before catching up (see [here](https://github.com/ava-labs/awm-relayer/blob/main/relayer/relayer.go#L121-L130)), I'd suggest that we leave that clamping logic in place, so that we don't doubly queue any blocks.